src/Cache/CacheSet.cpp

```diff
@@ -99,25 +97,31 @@ void NOMAD::CacheSet::destroy()
 void NOMAD::CacheSet::setInstance(const std::shared_ptr<NOMAD::CacheParameters>& cacheParams,
                                   const BBOutputTypeList& bbOutputType)
 {
+    bool called_twice = false;
 #ifdef _OPENMP
-    // Lock cache set before creating instance
-    omp_set_lock(&_cacheLock);
-#endif
-
+#pragma omp critical(initCacheLock)
+    {
+#endif // _OPENMP
     if ( _single == nullptr )
     {
+#ifdef _OPENMP
+        omp_init_lock(&_cacheLock);
```

Comment:
As-was, to create a `CacheSet`, the caller uses `setInstance()`, taking a `std::shared_ptr<NOMAD::CacheParameters>`. The `setInstance()` method then uses those parameters to create the instance, though `CacheSet`'s private constructor: ` _single = std::unique_ptr<NOMAD::CacheSet>(new CacheSet(cacheParams)) ;`. At last, the `init()` method is called in the body of said private constructor.

With OpenMP, first, `setInstance()` is entered. It calls `omp_set_lock(&_cacheLock);`, protecting `_single` to guarantee singleton creation of the `CacheSet`. Second, the `CacheSet` gets created as described above. Third, `_cacheLock` gets initialized in `init()`.

On Linux, which does not (apparently) allocate resources on initialization of the lock, the `omp_set_lock` apparently succeeds, but the above means that the subsequent `omp_init_lock(&_cacheLock);` will reset it, leaving the code open to a race condition and multiple `CacheSet` instances.

On Windows, which _does_ allocate resources on initialization of the lock, the `omp_set_lock()` on the uninitialized `_cacheLock` segfaults and the program aborts.

Regardless that it has the appearance of working on Linux, in all cases, a lock should be initialized before set.

Now, personally, I'd create all locks that are needed on program startup (i.e. when guaranteed serial), rather than on demand as is done here and in several other places. Point being that "on-demand" likely means "after threads have been started." Barring the former, and given the issues with `called_twice` and `_single` on top of the `_cacheLock`, I took the cop-out of a critical section to protect them all: it protects against initializing twice using `_single` as a predicate to enter the initialization of same and hence the lock initialization moved where it did.

This is not pretty, but since `setInstance()` is only ever supposed to be called once, the cost of the critical section isn't going to drag down application performance in any measurable way.